File: DSA/src/Questions/BinarySearch/MatrixSquareFinder.java
package Questions.BinarySearch;

public class MatrixSquareFinder {

    // Method to find the first and last occurrences of 1s
    public static int[][] findSquareOccurrences(int[][] matrix) {
        int m = matrix.length; // Number of rows
        int n = matrix[0].length; // Number of columns

        int firstRow = -1, firstCol = -1;
        int lastRow = -1, lastCol = -1;

        // Perform binary search to find the first occurrence of 1
        int low = 0, high = m * n - 1;

        // Find the first occurrence of 1
        while (low <= high) {
            int mid = (low + high) / 2;
            int row = mid / n; // Get row index
            int col = mid % n; // Get column index

            if (matrix[row][col] == 1) {
                firstRow = row;
                firstCol = col;
                high = mid - 1; // Search left for the first occurrence
            } else {
                low = mid + 1; // Search right
            }
        }

        // Reset low and high for last occurrence search
        low = 0;
        high = m * n - 1;

        // Find the last occurrence of 1
        while (low <= high) {
            int mid = (low + high) / 2;
            int row = mid / n; // Get row index
            int col = mid % n; // Get column index

            if (matrix[row][col] == 1) {
                lastRow = row;
                lastCol = col;
                low = mid + 1; // Search right for the last occurrence
            } else {
                high = mid - 1; // Search left
            }
        }

        return new int[][] { { firstRow, firstCol }, { lastRow, lastCol } };
    }

    public static void main(String[] args) {
        int[][] matrix = {
                { 0, 0, 0, 0 },
                { 0, 1, 1, 0 },
                { 0, 1, 1, 0 },
                { 0, 0, 0, 0 }
        };

        int[][] result = findSquareOccurrences(matrix);
        System.out.println("First occurrence of square of 1s: (" + result[0][0] + ", " + result[0][1] + ")");
        System.out.println("Last occurrence of square of 1s: (" + result[1][0] + ", " + result[1][1] + ")");
    }
}

File: DSA/Python/Questions/BinarySearch/MatrixSquareFinder.py
"""
# Question: Matrix Square Finder
# Problem Statement:
# Find the largest square submatrix containing all 1s in a binary matrix.

# Example:
# Input: matrix = [
#     [1,1,1],
#     [1,1,1],
#     [1,1,1]
# ]
# Output: 3
# Explanation: The entire matrix forms a 3x3 square of 1s
"""

from typing import List


class MatrixSquareFinder:
    def largest_square_binary(self, matrix: List[List[int]]) -> int:
        """
        Binary search approach
        Time Complexity: O(m*n*log(min(m,n)))
        Space Complexity: O(1)
        """
        if not matrix or not matrix[0]:
            return 0

        m, n = len(matrix), len(matrix[0])
        left, right = 0, min(m, n)

        while left < right:
            mid = (left + right + 1) >> 1
            if self.has_square(matrix, mid):
                left = mid
            else:
                right = mid - 1

        return left

    def has_square(self, matrix: List[List[int]], size: int) -> bool:
        """Helper function to check if square of given size exists"""
        m, n = len(matrix), len(matrix[0])
        for i in range(m - size + 1):
            for j in range(n - size + 1):
                if self.check_square(matrix, i, j, size):
                    return True
        return False

    def check_square(
        self, matrix: List[List[int]], row: int, col: int, size: int
    ) -> bool:
        """Helper function to verify if square at given position is valid"""
        for i in range(row, row + size):
            for j in range(col, col + size):
                if matrix[i][j] == 0:
                    return False
        return True

    def largest_square_dp(self, matrix: List[List[int]]) -> int:
        """
        Dynamic programming approach
        Time Complexity: O(m*n)
        Space Complexity: O(m*n)
        """
        if not matrix or not matrix[0]:
            return 0

        m, n = len(matrix), len(matrix[0])
        dp = [[0] * (n + 1) for _ in range(m + 1)]
        max_size = 0

        for i in range(1, m + 1):
            for j in range(1, n + 1):
                if matrix[i - 1][j - 1] == 1:
                    dp[i][j] = min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]) + 1
                    max_size = max(max_size, dp[i][j])

        return max_size

    def largest_square_optimized(self, matrix: List[List[int]]) -> int:
        """
        Space optimized approach using two rows
        Time Complexity: O(m*n)
        Space Complexity: O(n)
        """
        if not matrix or not matrix[0]:
            return 0

        m, n = len(matrix), len(matrix[0])
        prev = [0] * (n + 1)
        max_size = 0

        for i in range(1, m + 1):
            curr = [0] * (n + 1)
            for j in range(1, n + 1):
                if matrix[i - 1][j - 1] == 1:
                    curr[j] = min(prev[j], curr[j - 1], prev[j - 1]) + 1
                    max_size = max(max_size, curr[j])
            prev = curr

        return max_size


def main():
    test_cases = [
        {"matrix": [[1, 1, 1], [1, 1, 1], [1, 1, 1]], "expected": 3},
        {"matrix": [[1, 0, 1], [1, 1, 0], [1, 1, 0]], "expected": 1},
        {"matrix": [[0, 1], [1, 0]], "expected": 1},
    ]

    solution = MatrixSquareFinder()
    for i, test in enumerate(test_cases):
        print(f"\nTest Case {i + 1}:")
        print(f"Matrix: {test['matrix']}")

        result_binary = solution.largest_square_binary(test["matrix"])
        result_dp = solution.largest_square_dp(test["matrix"])
        result_optimized = solution.largest_square_optimized(test["matrix"])

        print(f"Binary Search Result: {result_binary}")
        print(f"DP Result: {result_dp}")
        print(f"Optimized Result: {result_optimized}")
        print(f"Expected: {test['expected']}")

        assert result_binary == result_dp == result_optimized == test["expected"]


if __name__ == "__main__":
    main()
